# 📐 코사인 유사도 (Cosine Similarity): 우리는 얼마나 닮았을까?

> **"너랑 나랑은 바라보는 방향이 같아!"**

사람끼리 "성격이 비슷하다"는 걸 어떻게 숫자로 나타낼 수 있을까?
텍스트 분석에서는 **코사인 유사도**라는 것을 가장 많이 사용해.

### 핵심 개념
- 두 문서(벡터)가 **비슷한 방향**을 가리킬수록 유사하다고 판단해.
- **값의 범위**: **0 ~ 1**
    - **1**: 완전히 똑같음 (일심동체!)
    - **0**: 전혀 상관없음 (남남)
    - (참고: -1은 정반대라는 뜻인데, 텍스트 분석에선 보통 0까지만 나와)

---

## 1. 간단한 예제로 이해하기
영화 '매트릭스'의 명대사를 가지고 실험을 해보자. 
3개의 문장이 있어.

1. **문장 A**: 파란 약을 먹으면 이야기는 끝난다.
2. **문장 B**: 빨간 약을 먹으면 원더랜드에 남는다.
3. **문장 C**: 빨간 약을 먹으면 토끼 굴 끝까지 보여주겠다.

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

# 3개의 문서 리스트
doc_list = [
    'if you take the blue pill, the story ends' ,                       # 문서 0
    'if you take the red pill, you stay in Wonderland',                 # 문서 1
    'if you take the red pill, I show you how deep the rabbit hole goes'# 문서 2
]

# 1. TF-IDF 벡터화 (글자 -> 숫자 변환)
tfidf_vect = TfidfVectorizer()
feature_vect = tfidf_vect.fit_transform(doc_list)

print("행렬 크기:", feature_vect.shape)
# (3개 문서, 단어 18개)

# 실제 벡터값 확인 (0이 아닌 값만 출력됨, 희소행렬)
# print(feature_vect)

행렬 크기: (3, 18)
<Compressed Sparse Row sparse matrix of dtype 'float64'
	with 12 stored elements and shape (1, 18)>
  Coords	Values
  (0, 6)	0.1830059506093466
  (0, 17)	0.3660119012186932
  (0, 14)	0.1830059506093466
  (0, 15)	0.3660119012186932
  (0, 8)	0.1830059506093466
  (0, 10)	0.23565348175165166
  (0, 11)	0.3098560092999078
  (0, 5)	0.3098560092999078
  (0, 1)	0.3098560092999078
  (0, 9)	0.3098560092999078
  (0, 4)	0.3098560092999078
  (0, 3)	0.3098560092999078


## 2. 유사도 계산하기 🧮

`cosine_similarity(X, Y)` 함수를 쓰면 X와 Y 사이의 유사도를 구해줘.
우리는 **(문서 0, 1) vs (문서 2)** 가 얼마나 닮았는지 비교해볼게.

In [3]:
# 문서 2 (rabbit hole...) 와 전체 문서 간의 유사도 측정
similarity = cosine_similarity(feature_vect[2], feature_vect)

print(similarity)

[[0.40425045 0.45647296 1.        ]]


### 🔍 결과 해석

결과가 `[[0.40..., 0.45..., 1. ]]` 이렇게 나왔다면:

1. **문서 2 vs 문서 0** (파란약): 유사도 **0.40**
2. **문서 2 vs 문서 1** (빨간약): 유사도 **0.45** (오! 빨간약끼리라서 더 높네!)
3. **문서 2 vs 문서 2** (자기 자신): 유사도 **1.0** (당연히 똑같지!)

-> 컴퓨터가 "파란 약" 문장보다 "빨간 약" 문장이랑 더 비슷하다고 판단했어. 아주 똑똑하지? 👍

---

## 🎓 마무리 퀴즈

**Q. 두 문서가 완전히 똑같은 내용이라면 코사인 유사도는 몇이 나올까?**
1. 0
2. 0.5
3. 1

<details>
<summary>정답 확인</summary>

**3번 (1)**

</details>